Add an initial WebKit User Agent Overrides file

sysrqb opened the pull request:
#### c8a200f9821c0925fb949defb10ca51d1f6cc34d
<pre>
Add an initial WebKit User Agent Overrides file
<a href="https://bugs.webkit.org/show_bug.cgi?id=263619">https://bugs.webkit.org/show_bug.cgi?id=263619</a>
<a href="https://rdar.apple.com/59289090">rdar://59289090</a>

Reviewed by NOBODY (OOPS!).

Create a new user agent string override configuration file. The configuration
file is JSON, and it is parsed and roughly validated at compile-time as a
sanity check. At run-time the JSON file is read in the UI process where it is
re-parsed and re-validated. This additional step allows for testing new user
agent quirks at runtime without needing to recompile WebKit.

The behavior is currently somewhat minimal, but it can easily be extended and
enhanced depending on our needs.

Version 1 input JSON file contains an object with:
  - a version, currently the only supported value is the integer 1.
  - objects where the key is a domain:
    - objects where the key is either the target platform (e.g., &quot;mac&quot;) or &quot;all&quot;, and the value is either:
      - an array of attributes that should be used in the overridden string:
        - user agent (e.g., chrome)
        - platform (e.g., mac; the current platform is the default, if not specified)
      - a string specifying a hard-coded user agent string

For example, overriding iOS with the Mac Chrome user agent string, and Mac with Chrome&apos;s Mac user agent string could look like:
  &quot;target.site.example&quot;: { &quot;ios&quot;: [ &quot;chrome&quot;, &quot;mac&quot; ], &quot;mac&quot;: [ &quot;chrome&quot; ] }

* Source/WebCore/CMakeLists.txt:
* Source/WebCore/DerivedSources-input.xcfilelist:
* Source/WebCore/DerivedSources.make:
* Source/WebCore/Headers.cmake:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/WebCoreMacros.cmake:
* Source/WebCore/loader/FrameLoader.cpp:
(WebCore::FrameLoader::userAgent const):
* Source/WebCore/page/Quirks.cpp:
(WebCore::updatableStorageAccessUserAgentStringQuirks): Deleted.
(WebCore::Quirks::updateStorageAccessUserAgentStringQuirks): Deleted.
(WebCore::Quirks::storageAccessUserAgentStringQuirkForDomain): Deleted.
* Source/WebCore/page/Quirks.h:
* Source/WebCore/page/UserAgentStringOverrides.cpp: Added.
(WebCore::UserAgentStringOverrides::getPlatform):
(WebCore::UserAgentStringOverrides::stringToUserAgentPlatform):
(WebCore::UserAgentStringOverrides::stringToUserAgent):
(WebCore::UserAgentStringOverrides::userAgentAttributeToString):
(WebCore::platformToNavigatorPlatformString):
(WebCore::platformToPlatformString):
(WebCore::agentToAgentVersionString):
(WebCore::agentToAgentString):
(WebCore::UserAgentStringOverrides::attributesToPlatformString):
(WebCore::UserAgentStringOverrides::attributesToNavigatorPlatformString):
(WebCore::UserAgentStringOverrides::attributesToAgentString):
(WebCore::UserAgentStringOverrides::attributesToPlatformVersionString):
(WebCore::UserAgentStringOverrides::attributesToAgentVersionString):
(WebCore::UserAgentStringOverrides::userAgentToString):
(WebCore::UserAgentStringOverrides::parseUserAgentOverrides):
(WebCore::UserAgentStringOverrides::attributesToPlatform):
(WebCore::UserAgentStringOverrides::attributesToAgent):
(WebCore::UserAgentStringOverrides::setUserAgentStringQuirks):
(WebCore::UserAgentStringOverrides::overrideAttributesForURLAndType const):
* Source/WebCore/page/UserAgentStringOverrides.h: Added.
* Source/WebCore/page/UserAgentStringOverrides.json: Added.
* Source/WebCore/page/make-user-agent-string-overrides.py: Added.
(main):
* Source/WebKit/UIProcess/Cocoa/WebProcessPoolCocoa.mm:
(WebKit::WebProcessPool::userAgentOverrideDirectory):
(WebKit::WebProcessPool::additionalUserAgentOverrideDirectoryForTesting):
(WebKit::WebProcessPool::platformInitializeWebProcess):
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::decidePolicyForNavigationAction):
(WebKit::WebPageProxy::adjustPolicyWithCustomUserAgentAndPlatform):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/WebProcessPool.cpp:
(WebKit::WebProcessPool::userAgentOverrideDirectory):
(WebKit::WebProcessPool::additionalUserAgentOverrideDirectory):
(WebKit::readUserAgentStringOverridesFromFile):
(WebKit::WebProcessPool::setUserAgentStringQuirks):
(WebKit::WebProcessPool::userAgentStringQuirks const):
* Source/WebKit/UIProcess/WebProcessPool.h:
* Source/WebKit/UIProcess/ios/WebPageProxyIOS.mm:
(WebKit::WebPageProxy::effectiveContentModeAfterAdjustingPolicies):
* Source/WebKit/WebProcess/WebProcess.cpp:
(WebKit::WebProcess::initializeWebProcess):
(WebKit::WebProcess::updateStorageAccessUserAgentStringQuirks): Deleted.
* Source/WebKit/WebProcess/WebProcess.h:
* Source/WebKit/WebProcess/WebProcess.messages.in:
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebKit/UserAgentStringOverrides.json: Added.
* Tools/TestWebKitAPI/Tests/WebKitCocoa/Navigation.mm:
(TEST(WKNavigation, UserAgentOverrideQuirks)):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/ResourceLoadStatistics.mm:
(TEST(ResourceLoadStatistics, UserAgentStringForSite)): Deleted.
</pre><!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c8a200f9821c0925fb949defb10ca51d1f6cc34d

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/72435 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/51856 "Built successfully") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/25223 "Build is in progress. Recent messages:Running configuration; Running apply-patch; Checked out pull request; Compiled WebKit (warnings)") | [❌ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/76616 "Hash c8a200f9 for PR 29163 does not build (failure)") | [❌ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/23644 "Hash c8a200f9 for PR 29163 does not build (failure)") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/74550 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/59660 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/23466 "Built successfully") | [❌ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/5/builds/76616 "Hash c8a200f9 for PR 29163 does not build (failure)") | [❌ 🧪 win-tests](https://ews-build.webkit.org/#/builders/59/builds/23644 "Hash c8a200f9 for PR 29163 does not build (failure)") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/75502 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/46951 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/62382 "Passed tests") | [❌ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/5/builds/76616 "Hash c8a200f9 for PR 29163 does not build (failure)") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/43602 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/19848 "Passed tests") | [❌ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/21994 "Hash c8a200f9 for PR 29163 does not build (failure)") | 
| | [❌ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/65483 "34 api tests failed or timed out") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/20206 "Passed tests") | [❌ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/78283 "Failed to compile WebKit") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/16676 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/19340 "Passed tests") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/2/builds/78283 "Failed to compile WebKit") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/16724 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/62394 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/2/builds/78283 "Failed to compile WebKit") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/13033 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/6673 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/47654 "Built successfully") | [❌ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/2439 "Found 1 new failure in page/UserAgentStringOverrides.cpp") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/48723 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/50012 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/48466 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

karlcow commented on Source/WebCore/page/make-user-agent-string-quirks.py:
```diff
@@ -0,0 +1,178 @@
+            output_file.write(text)
+
+        def writeln(text):
+            write(text + '\n')
```
Was there a performance decision for passing by this function to generate the `.cpp` file?

sysrqb replied:
Mostly compile-time validation and ease of implementation.

karlcow commented on Source/WebCore/page/make-user-agent-string-quirks.py:
```diff
@@ -0,0 +1,178 @@
+        for platform in sorted(user_agents.keys()):
+            for user_agent in user_agents[platform]:
+                entry = platform.upper() + "_" + user_agent.upper() + ","
+            writeln("        " + entry)
```
This makes me a bit nervous. It should probably be a function. And then we could create a unit test to make double sure, only the right format is generated.

karlcow commented on Source/WebCore/page/make-user-agent-string-quirks.py:
```diff
@@ -0,0 +1,178 @@
+                    raise Exception("Unknown platform: {}".format(platform))
+                user_agent = overrides[domain][platform]
+                if user_agent not in user_agents[platform]:
+                    raise Exception("Unknown user agent string for: {}_{}".format(platform, user_agent))
```
similar comment than above.

karlcow commented on Source/WebCore/page/make-user-agent-string-quirks.py:
```diff
@@ -0,0 +1,178 @@
+
+
+def lowercase_first_letter(string):
+    return string[0].lower() + string[1:] if len(string) and string[1].islower() else string
```
nit: This function is not used in the rest of the code. Breadcrumb leftover?

sysrqb replied:
Yep, not needed anymore.

Commit:
Add an initial WebKit User Agent Overrides file
https://bugs.webkit.org/show_bug.cgi?id=263619
rdar://59289090

Reviewed by NOBODY (OOPS!).

Create a new user agent string override configuration file. The configuration
file is JSON, and it is parsed and roughly validated at compile-time as a
sanity check. At run-time the JSON file is read in the UI process where it is
re-parsed and re-validated. This additional step allows for testing new user
agent quirks at runtime without needing to recompile WebKit.

The behavior is currently somewhat minimal, but it can easily be extended and
enhanced depending on our needs.

Version 1 input JSON file contains an object with:
  - a version, currently the only supported value is the integer 1.
  - objects where the key is a domain:
    - objects where the key is either the target platform (e.g., "mac") or "all", and the value is either:
      - an array of attributes that should be used in the overridden string:
        - user agent (e.g., chrome)
        - platform (e.g., mac; the current platform is the default, if not specified)
      - a string specifying a hard-coded user agent string

For example, overriding iOS with the Mac Chrome user agent string, and Mac with Chrome's Mac user agent string could look like:
  "target.site.example": { "ios": [ "chrome", "mac" ], "mac": [ "chrome" ] }

* Source/WebCore/CMakeLists.txt:
* Source/WebCore/DerivedSources-input.xcfilelist:
* Source/WebCore/DerivedSources.make:
* Source/WebCore/Headers.cmake:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/WebCoreMacros.cmake:
* Source/WebCore/loader/FrameLoader.cpp:
(WebCore::FrameLoader::userAgent const):
* Source/WebCore/page/Quirks.cpp:
(WebCore::updatableStorageAccessUserAgentStringQuirks): Deleted.
(WebCore::Quirks::updateStorageAccessUserAgentStringQuirks): Deleted.
(WebCore::Quirks::storageAccessUserAgentStringQuirkForDomain): Deleted.
* Source/WebCore/page/Quirks.h:
* Source/WebCore/page/UserAgentStringOverrides.cpp: Added.
(WebCore::UserAgentStringOverrides::getPlatform):
(WebCore::UserAgentStringOverrides::stringToUserAgentPlatform):
(WebCore::UserAgentStringOverrides::stringToUserAgent):
(WebCore::UserAgentStringOverrides::userAgentAttributeToString):
(WebCore::platformToNavigatorPlatformString):
(WebCore::platformToPlatformString):
(WebCore::agentToAgentVersionString):
(WebCore::agentToAgentString):
(WebCore::UserAgentStringOverrides::attributesToPlatformString):
(WebCore::UserAgentStringOverrides::attributesToNavigatorPlatformString):
(WebCore::UserAgentStringOverrides::attributesToAgentString):
(WebCore::UserAgentStringOverrides::attributesToPlatformVersionString):
(WebCore::UserAgentStringOverrides::attributesToAgentVersionString):
(WebCore::UserAgentStringOverrides::userAgentToString):
(WebCore::UserAgentStringOverrides::parseUserAgentOverrides):
(WebCore::UserAgentStringOverrides::attributesToPlatform):
(WebCore::UserAgentStringOverrides::attributesToAgent):
(WebCore::UserAgentStringOverrides::setUserAgentStringQuirks):
(WebCore::UserAgentStringOverrides::overrideAttributesForURLAndType const):
* Source/WebCore/page/UserAgentStringOverrides.h: Added.
* Source/WebCore/page/UserAgentStringOverrides.json: Added.
* Source/WebCore/page/make-user-agent-string-overrides.py: Added.
(main):
* Source/WebKit/UIProcess/Cocoa/WebProcessPoolCocoa.mm:
(WebKit::WebProcessPool::userAgentOverrideDirectory):
(WebKit::WebProcessPool::additionalUserAgentOverrideDirectoryForTesting):
(WebKit::WebProcessPool::platformInitializeWebProcess):
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::decidePolicyForNavigationAction):
(WebKit::WebPageProxy::adjustPolicyWithCustomUserAgentAndPlatform):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/WebProcessPool.cpp:
(WebKit::WebProcessPool::userAgentOverrideDirectory):
(WebKit::WebProcessPool::additionalUserAgentOverrideDirectory):
(WebKit::readUserAgentStringOverridesFromFile):
(WebKit::WebProcessPool::setUserAgentStringQuirks):
(WebKit::WebProcessPool::userAgentStringQuirks const):
* Source/WebKit/UIProcess/WebProcessPool.h:
* Source/WebKit/UIProcess/ios/WebPageProxyIOS.mm:
(WebKit::WebPageProxy::effectiveContentModeAfterAdjustingPolicies):
* Source/WebKit/WebProcess/WebProcess.cpp:
(WebKit::WebProcess::initializeWebProcess):
(WebKit::WebProcess::updateStorageAccessUserAgentStringQuirks): Deleted.
* Source/WebKit/WebProcess/WebProcess.h:
* Source/WebKit/WebProcess/WebProcess.messages.in:
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebKit/UserAgentStringOverrides.json: Added.
* Tools/TestWebKitAPI/Tests/WebKitCocoa/Navigation.mm:
(TEST(WKNavigation, UserAgentOverrideQuirks)):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/ResourceLoadStatistics.mm:
(TEST(ResourceLoadStatistics, UserAgentStringForSite)): Deleted.

mcatanzaro commented on Source/WebCore/Headers.cmake:
```diff
@@ -2752,6 +2753,7 @@ set(WebCore_PRIVATE_FRAMEWORK_HEADERS
     ${WebCore_DERIVED_SOURCES_DIR}/StreamInternalsBuiltins.h
     ${WebCore_DERIVED_SOURCES_DIR}/TagName.h
     ${WebCore_DERIVED_SOURCES_DIR}/TransformStreamInternalsBuiltins.h
+    ${WebCore_DERIVED_SOURCES_DIR}/UserAgentStringOverrides.h
     ${WebCore_DERIVED_SOURCES_DIR}/UserAgentParts.h
```
This isn't alphabetized properly.

mcatanzaro commented on Source/WebCore/WebCoreMacros.cmake:
```diff
@@ -220,3 +220,14 @@ function(GENERATE_DOM_NAME_ENUM _enum)
+    set(OVERRIDES_VALIDATION_SCRIPT ${WEBCORE_DIR}/page/make-user-agent-string-overrides.py)
+
+    add_custom_command(
+        TARGET WebCore PRE_LINK
```
I think we can remove this line, right? The json file can be validated at any time? No need to delay until after all of WebCore has already been compiled. If there's an error in the json, it's good for the build to fail sooner rather than later.

mcatanzaro commented on Source/WebKit/UIProcess/WebProcessPool.cpp:
```diff
@@ -2363,6 +2389,19 @@ void WebProcessPool::setUseSeparateServiceWorkerProcess(bool useSeparateServiceW
         processPool->terminateServiceWorkers();
 }
 
+template<typename T>
+void WebProcessPool::setUserAgentStringQuirks(T overrides)
```
Why is this a template?

mcatanzaro commented on Source/WebCore/DerivedSources.make:
```diff
@@ -2368,6 +2368,18 @@ $(EVENT_NAME_FILES_PATTERNS) : $(WebCore)/dom/make-event-names.py $(EVENT_NAME_J
+#
+
+all : $(USER_AGENT_STRING_OVERRIDE_JSON)
+$(USER_AGENT_STRING_OVERRIDE_FILES_PATTERNS) : $(WebCore)/page/make-user-agent-string-overrides.py $(USER_AGENT_STRING_OVERRIDE_JSON)
+	$(PYTHON) "$(WebCore)/page/make-user-agent-string-overrides.py" --input $(USER_AGENT_STRING_OVERRIDE_JSON)
```
This script needs to be renamed since it's no longer making anything, only validating the json. (Which was a good change.)

mcatanzaro commented on Source/WebCore/page/make-user-agent-string-overrides.py:
```diff
@@ -0,0 +1,84 @@
+    # Version 1 input JSON file contains an object with:
+    #   - a version, currently the only supported value is the integer 1.
+    #   - objects where the key is a domain:
+    #     - objects where the key is either the target platform (e.g., "macos") or "all", and the value is either:
+    #       - an array of attributes that should be used in the overridden string:
+    #         - user agent (e.g., chrome)
+    #         - platform (e.g., mac)
```
We'll also need to be able to perform a chassis type check (currently "desktop" or "mobile").

mcatanzaro commented on Source/WebCore/page/make-user-agent-string-overrides.py:
```diff
@@ -0,0 +1,84 @@
+    parser.add_argument('--input')
+    args = parser.parse_args()
+
+    # Version 1 input JSON file contains an object with:
+    #   - a version, currently the only supported value is the integer 1.
+    #   - objects where the key is a domain:
```
We're going to need a way to specify if a quirk should apply to (a) the domain, exactly as specified, or (b) the toplevel base domain. E.g. one quirk for google.com might apply only to google.com and not also to drive.google.com, but another might need to apply to all subdomains of google.com.

mcatanzaro reviewed (commented):
OK, I think this is on the right track. It needs more work to pass EWS, and I'll need to follow up to integrate it for WPE/GTK and try to migrate the existing quirks. But all my previous concerns are clearly resolved. Thanks.
